File: ios/TimberCruisingApp/Sensors/SamplingStats.swift
// Sampling-statistics engine — pure functions for cruise design and
// post-cruise reporting. Adopted from SilvaCruise's "full sampling-
// statistics engine" pitch (CV / SE / CI / SDI / Reineke / Curtis /
// VBAR / Neyman). All inputs are plain Swift arrays of plot-level
// metrics; nothing here touches Core Data, AR, or persistence.
//
// What's covered:
//
//   • cv / se / mean / standardDeviation
//   • confidenceInterval(t: cv: n:) — two-sided
//   • requiredSampleSize(targetSEPct: cv: t:) — for cruise design
//   • neymanAllocation(strataCV:strataN:totalSampleSize:) — for
//     stratified cruise design
//   • reineke SDI helpers
//   • curtisRD — Curtis Relative Density (PNW standard)
//
// Tests in CommonTests can pin known values from cruise-statistics
// textbooks (Avery & Burkhart, Iles "Sampling Methods" reference).

import Foundation

public enum SamplingStats {

    // MARK: - Basic descriptive stats

    public static func mean(_ xs: [Double]) -> Double {
        guard !xs.isEmpty else { return 0 }
        return xs.reduce(0, +) / Double(xs.count)
    }

    /// Unbiased sample standard deviation (n − 1 denominator).
    /// Returns 0 for sample size < 2.
    public static func standardDeviation(_ xs: [Double]) -> Double {
        guard xs.count >= 2 else { return 0 }
        let m = mean(xs)
        let sumSq = xs.reduce(0) { $0 + ($1 - m) * ($1 - m) }
        return (sumSq / Double(xs.count - 1)).squareRoot()
    }

    /// Coefficient of variation as a percentage.
    public static func cv(_ xs: [Double]) -> Double {
        let m = mean(xs)
        guard m != 0 else { return 0 }
        return standardDeviation(xs) / m * 100
    }

    /// Standard error of the mean (absolute units, not %).
    public static func se(_ xs: [Double]) -> Double {
        guard xs.count >= 2 else { return 0 }
        return standardDeviation(xs) / Double(xs.count).squareRoot()
    }

    /// Standard error of the mean as a percentage of the mean (one-
    /// sided SE%, often called sampling error in cruise reports).
    public static func sePct(_ xs: [Double]) -> Double {
        let m = mean(xs)
        guard m != 0 else { return 0 }
        return se(xs) / m * 100
    }

    // MARK: - Confidence intervals

    /// 95 % two-sided CI half-width on the mean using a Student-t
    /// approximation. `tApprox` defaults to 2.0 (n ≥ 30 rule of
    /// thumb); cruise designs that demand tight stats should pass
    /// the actual t-value for their df.
    public static func ciHalfWidth(_ xs: [Double], tApprox: Double = 2.0) -> Double {
        tApprox * se(xs)
    }

    public static func confidenceInterval(_ xs: [Double], tApprox: Double = 2.0)
        -> (lower: Double, upper: Double)? {
        guard xs.count >= 2 else { return nil }
        let m = mean(xs)
        let h = ciHalfWidth(xs, tApprox: tApprox)
        return (m - h, m + h)
    }

    // MARK: - Required sample size

    /// Target sample size for a desired one-sided sampling error
    /// (`targetSEPct`) given an estimated CV (`cv`, in %) and a
    /// t-multiplier (~2 for n ≥ 30, ~3 for very small n).
    /// Result rounds up.
    public static func requiredSampleSize(targetSEPct: Double,
                                          cv: Double,
                                          t: Double = 2.0) -> Int {
        guard targetSEPct > 0 else { return 0 }
        let n = pow((t * cv) / targetSEPct, 2)
        return Int(ceil(n))
    }

    // MARK: - Neyman allocation

    /// Optimal stratified sample-size allocation. For each stratum:
    ///     n_h = totalN × (W_h × σ_h) / Σ (W_k × σ_k)
    /// where `W_h = N_h / Σ N_k` (population weight). Inputs:
    ///   • `strataCV` — CV per stratum (%)
    ///   • `strataN` — population size per stratum (e.g. acres)
    ///   • `totalSampleSize` — total plots to allocate
    /// Returns plots per stratum, rounded down with the remainder
    /// distributed largest-fraction-first so rounding loss doesn't
    /// reduce the cruise's effective sample size.
    public static func neymanAllocation(
        strataCV: [Double],
        strataN: [Double],
        totalSampleSize: Int
    ) -> [Int] {
        guard strataCV.count == strataN.count, !strataCV.isEmpty,
              totalSampleSize > 0 else {
            return Array(repeating: 0, count: strataCV.count)
        }
        let totalN = strataN.reduce(0, +)
        guard totalN > 0 else { return Array(repeating: 0, count: strataCV.count) }
        // Use σ * W_h ∝ CV × N_h (the constant-mean factor cancels
        // out of the fraction; this is fine for typical cruise data).
        let weights = zip(strataCV, strataN).map { $0 * $1 }
        let sumW = weights.reduce(0, +)
        guard sumW > 0 else { return Array(repeating: 0, count: strataCV.count) }
        let raw = weights.map { Double(totalSampleSize) * $0 / sumW }
        var floored = raw.map(Int.init)   // floor
        var remaining = totalSampleSize - floored.reduce(0, +)
        let fractionsSorted = raw.enumerated()
            .map { ($0.offset, $0.element - Double(Int($0.element))) }
            .sorted { $0.1 > $1.1 }
        for (idx, _) in fractionsSorted where remaining > 0 {
            floored[idx] += 1
            remaining -= 1
        }
        return floored
    }

    // MARK: - Density indices

    /// Reineke SDI = TPA × (QMD_in / 10)^1.605
    /// Returns 0 for non-positive inputs.
    public static func reinekeSDI(tpa: Double, qmdInches: Double) -> Double {
        guard tpa > 0, qmdInches > 0 else { return 0 }
        return tpa * pow(qmdInches / 10.0, 1.605)
    }

    /// Reineke relative density vs a species Max SDI baseline,
    /// expressed as a percentage. Generic Max SDI of 717 is used
    /// downstream when species-specific tables aren't available.
    public static func reinekeRelativeDensityPct(sdi: Double, maxSDI: Double) -> Double {
        guard maxSDI > 0 else { return 0 }
        return min(200, max(0, sdi / maxSDI * 100))
    }

    /// Curtis Relative Density (PNW standard):
    ///     RD = BA / sqrt(QMD)   (BA in ft²/ac, QMD in inches)
    public static func curtisRD(baPerAcreFt2: Double, qmdInches: Double) -> Double {
        guard qmdInches > 0 else { return 0 }
        return baPerAcreFt2 / qmdInches.squareRoot()
    }

    // MARK: - Cruise rating

    /// Three-band cruise rating off the standard sampling-error
    /// percentage. These thresholds match the ones SilvaCruise uses
    /// for its "Acceptable / Marginal / Poor" verdict.
    public enum CruiseRating: String, Sendable {
        case acceptable, marginal, poor
    }

    public static func rating(forSEPct sePct: Double) -> CruiseRating {
        switch sePct {
        case ..<10:  return .acceptable
        case ..<20:  return .marginal
        default:     return .poor
        }
    }
}

File: android/app/src/main/java/com/hcjeong/forestix/sensors/SamplingStats.kt
// Sampling-statistics engine — pure functions for cruise design and
// post-cruise reporting. Direct port of iOS Sensors/SamplingStats.swift.
//
// What's covered:
//   • cv / se / mean / standardDeviation
//   • confidenceInterval — two-sided
//   • requiredSampleSize(targetSEPct: cv: t:) — for cruise design
//   • neymanAllocation — for stratified cruise design
//   • reineke SDI helpers
//   • curtisRD — Curtis Relative Density (PNW standard)
//
// All inputs are plain lists of plot-level metrics; nothing here touches
// Room, AR, or persistence.

package com.hcjeong.forestix.sensors

import kotlin.math.ceil
import kotlin.math.max
import kotlin.math.min
import kotlin.math.pow
import kotlin.math.sqrt

object SamplingStats {

    // MARK: - Basic descriptive stats

    fun mean(xs: List<Double>): Double {
        if (xs.isEmpty()) return 0.0
        return xs.sum() / xs.size.toDouble()
    }

    /// Unbiased sample standard deviation (n − 1 denominator).
    /// Returns 0 for sample size < 2.
    fun standardDeviation(xs: List<Double>): Double {
        if (xs.size < 2) return 0.0
        val m = mean(xs)
        val sumSq = xs.fold(0.0) { acc, x -> acc + (x - m) * (x - m) }
        return sqrt(sumSq / (xs.size - 1).toDouble())
    }

    /// Coefficient of variation as a percentage.
    fun cv(xs: List<Double>): Double {
        val m = mean(xs)
        if (m == 0.0) return 0.0
        return standardDeviation(xs) / m * 100.0
    }

    /// Standard error of the mean (absolute units, not %).
    fun se(xs: List<Double>): Double {
        if (xs.size < 2) return 0.0
        return standardDeviation(xs) / sqrt(xs.size.toDouble())
    }

    /// Standard error of the mean as a percentage of the mean (one-
    /// sided SE%, often called sampling error in cruise reports).
    fun sePct(xs: List<Double>): Double {
        val m = mean(xs)
        if (m == 0.0) return 0.0
        return se(xs) / m * 100.0
    }

    // MARK: - Confidence intervals

    /// 95 % two-sided CI half-width on the mean using a Student-t
    /// approximation. `tApprox` defaults to 2.0 (n ≥ 30 rule of thumb).
    fun ciHalfWidth(xs: List<Double>, tApprox: Double = 2.0): Double =
        tApprox * se(xs)

    fun confidenceInterval(xs: List<Double>, tApprox: Double = 2.0): Pair<Double, Double>? {
        if (xs.size < 2) return null
        val m = mean(xs)
        val h = ciHalfWidth(xs, tApprox = tApprox)
        return Pair(m - h, m + h)
    }

    // MARK: - Required sample size

    /// Target sample size for a desired one-sided sampling error
    /// (`targetSEPct`) given an estimated CV (`cv`, in %) and a
    /// t-multiplier (~2 for n ≥ 30, ~3 for very small n). Result rounds up.
    fun requiredSampleSize(targetSEPct: Double, cv: Double, t: Double = 2.0): Int {
        if (targetSEPct <= 0.0) return 0
        val n = ((t * cv) / targetSEPct).pow(2)
        return ceil(n).toInt()
    }

    // MARK: - Neyman allocation

    /// Optimal stratified sample-size allocation. For each stratum:
    ///     n_h = totalN × (W_h × σ_h) / Σ (W_k × σ_k)
    /// Returns plots per stratum, rounded down with the remainder
    /// distributed largest-fraction-first.
    fun neymanAllocation(
        strataCV: List<Double>,
        strataN: List<Double>,
        totalSampleSize: Int,
    ): List<Int> {
        if (strataCV.size != strataN.size || strataCV.isEmpty() || totalSampleSize <= 0) {
            return List(strataCV.size) { 0 }
        }
        val totalN = strataN.sum()
        if (totalN <= 0.0) return List(strataCV.size) { 0 }
        val weights = strataCV.zip(strataN).map { (c, n) -> c * n }
        val sumW = weights.sum()
        if (sumW <= 0.0) return List(strataCV.size) { 0 }
        val raw = weights.map { totalSampleSize.toDouble() * it / sumW }
        val floored = raw.map { it.toInt() }.toMutableList()
        var remaining = totalSampleSize - floored.sum()
        val fractionsSorted = raw.mapIndexed { idx, v -> Pair(idx, v - v.toInt().toDouble()) }
            .sortedByDescending { it.second }
        for ((idx, _) in fractionsSorted) {
            if (remaining <= 0) break
            floored[idx] += 1
            remaining -= 1
        }
        return floored
    }

    // MARK: - Density indices

    /// Reineke SDI = TPA × (QMD_in / 10)^1.605. Returns 0 for non-positive inputs.
    fun reinekeSDI(tpa: Double, qmdInches: Double): Double {
        if (tpa <= 0.0 || qmdInches <= 0.0) return 0.0
        return tpa * (qmdInches / 10.0).pow(1.605)
    }

    /// Reineke relative density vs a species Max SDI baseline, as a
    /// percentage. Generic Max SDI of 717 is used downstream when
    /// species-specific tables aren't available.
    fun reinekeRelativeDensityPct(sdi: Double, maxSDI: Double): Double {
        if (maxSDI <= 0.0) return 0.0
        return min(200.0, max(0.0, sdi / maxSDI * 100.0))
    }

    /// Curtis Relative Density (PNW standard):
    ///     RD = BA / sqrt(QMD)   (BA in ft²/ac, QMD in inches)
    fun curtisRD(baPerAcreFt2: Double, qmdInches: Double): Double {
        if (qmdInches <= 0.0) return 0.0
        return baPerAcreFt2 / sqrt(qmdInches)
    }

    // MARK: - Cruise rating

    /// Three-band cruise rating off the standard sampling-error percentage.
    /// Raw values match the Swift enum cases byte-for-byte.
    enum class CruiseRating(val raw: String) {
        ACCEPTABLE("acceptable"),
        MARGINAL("marginal"),
        POOR("poor"),
    }

    fun rating(forSEPct: Double): CruiseRating = when {
        forSEPct < 10.0 -> CruiseRating.ACCEPTABLE
        forSEPct < 20.0 -> CruiseRating.MARGINAL
        else -> CruiseRating.POOR
    }
}
